ramips: improve support for MSG1500 X.00

aiamadeus opened the pull request:
The Config partition of some machines is special, and the openwrt script cannot read the protest_lan_mac correctly (like #9468). This problem can be solved by reading the mac address (ascii) in dts. Also add another alternative name for this machine.

DragonBluep commented on target/linux/ramips/dts/mt7621_raisecom_msg1500-x-00.dts:
```diff
@@ -150,13 +176,3 @@
 };
-
-&factory {
-	compatible = "nvmem-cells";
```
I think this node can be removed because the mt76 driver always automatically get the WLAN MAC from `factory@0x04`.

AmadeusGhost replied:
Without this, mt76 will read factory@0xa as radio1's mac address. And https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac#L85 only fixes the mac address of radio0.
So this `nvmem-cells = <&macaddr_factory_4>;` can't be removed.

DragonBluep commented on target/linux/ramips/dts/mt7621_raisecom_msg1500-x-00.dts:
```diff
@@ -79,12 +80,30 @@
+			#size-cells = <1>;
+
+			macaddr_config_8014: macaddr@8014 {
+				reg = <0x8014 0x11>;
```
I can confirm that the offset is correct.

DragonBluep commented:
Hey, did you get the correct lan/wan mac, it looks like mac-address-ascii doesn't work properly on mt7621 or nand flash.

Ansuel commented:
@AmadeusGhost @DragonBluep reported wrong mac can we check this?

DragonBluep commented:
The patch for the kernel 5.15 needs to be refreshed. It works fine on the kernel 5.10. Unfortunately I sent all of my devices back home so I have no way to verify it.

Ansuel commented:
@aiamadeus can you refresh this?

aiamadeus commented:
> can you refresh this?

Yes, but my msg1500 has been burned and I can't test it.

aiamadeus committed:
ramips: MSG1500 X.00: switch to mac-address-ascii dts

The Config partition of some machines is special, and the openwrt script
cannot read the protest_lan_mac correctly. This problem can be solved by
reading the mac address (ascii) in dts.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

Ansuel commented:
@aiamadeus burned o.o?

DragonBluep commented:
Tested on another mt7621 router and confirmed that `mac-address-ascii` does not work on mt7531 switch.

Ansuel commented:
@DragonBluep so this doesn't work ?

DragonBluep commented:
@Ansuel Sadly it seems not. But first two commits are well and useful. Is there any way to dump the data of nvmem cell?

hauke commented:
You can access the mtd partitions directly, they are located in ''/dev/mtd*'' The mapping of the names is in `cat /proc/mtd`.

hauke commented:
Thank you for the patches, I applied the first 2 patches to master in 7b863af180dac48ea392a5ee582e37d0f378b3d3 and 4f9b360f0b9a85202422ef07ee573eeca06d11ab.

Ansuel commented:
@DragonBluep can you give this another spin? I would love to get this merged

DragonBluep commented:
@Ansuel It still doesn't work after rebased to current master branch. The MT7530 switch seems to be incompatible with `mac-address-ascii` NVMEM cell, which was also confirmed by @csharper2005 in https://github.com/openwrt/openwrt/pull/11712#discussion_r1066641386.
```
# These errors shouled be harmless.
Sat Jan 21 15:06:21 2023 kern.err kernel: [    0.606269] OF: /ethernet@1e100000/mac@0: could not get #nvmem-cell-cells for /nand@1e003000/partitions/partition@80000/macaddr@8014
Sat Jan 21 15:06:21 2023 kern.err kernel: [    0.629922] OF: /ethernet@1e100000/mac@1: could not get #nvmem-cell-cells for /nand@1e003000/partitions/partition@80000/macaddr@8036
Sat Jan 21 15:06:21 2023 kern.err kernel: [    0.654187] OF: /pcie@1e140000/pcie@0,0/wifi@0,0: could not get #nvmem-cell-cells for /nand@1e003000/partitions/partition@100000/macaddr@4
# still random MAC address 
Sat Jan 21 15:06:21 2023 kern.err kernel: [    2.564492] mtk_soc_eth 1e100000.ethernet: generated random MAC address be:5c:2b:ec:97:0c
Sat Jan 21 15:06:21 2023 kern.info kernel: [    2.583778] mtk_soc_eth 1e100000.ethernet eth0: mediatek frame engine at 0xbe100000, irq 21
Sat Jan 21 15:06:21 2023 kern.err kernel: [    2.600702] mtk_soc_eth 1e100000.ethernet: generated random MAC address 8a:26:46:1a:3c:05
Sat Jan 21 15:06:21 2023 kern.info kernel: [    2.619950] mtk_soc_eth 1e100000.ethernet wan: mediatek frame engine at 0xbe100000, irq 21
```
And I also tried u-boot env NVMEM on H3C TX180x series 13308161788c98ae6cd48c22b13339fdb8c77130, just like the commits you pushed 952c738c0fe360ae074fbf6a87d03883157bc2dd 5dee5965012e788f06e4d095e8cfb73200d818cb. Sadly, I got the same result.

BTW, this part should be removed if it can work one day.
https://github.com/openwrt/openwrt/blob/6e3acebe35cfc584e145e495b175ae5308d2f2bc/target/linux/ramips/mt7621/base-files/etc/board.d/02_network#L276-L280

DragonBluep commented:
@Ansuel Finally, I know why it doesn't work. This is because it does not call `nvmem_get_mac_address()` at all since mt7530 is not a platform device.

I added this simple patch to test:
```
--- a/net/core/of_net.c
+++ b/net/core/of_net.c
@@ -71,9 +71,11 @@ static int of_get_mac_addr_nvmem(struct
 	if (pdev) {
 		ret = nvmem_get_mac_address(&pdev->dev, addr);
 		put_device(&pdev->dev);
+		printk(KERN_ERR "It's a pdev\n");
 		return ret;
 	}
 
+	printk(KERN_ERR "It's not a pdev\n");
 	cell = of_nvmem_cell_get(np, "mac-address");
 	if (IS_ERR(cell))
 		return PTR_ERR(cell);
```
And then I got the result: 
```
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.387772] mt7530 mdio-bus:1f: MT7530 adapts as multi-chip module
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.403968] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.409613] mtk_soc_eth 1e100000.ethernet: generated random MAC address c6:d6:fb:05:33:23
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.428990] mtk_soc_eth 1e100000.ethernet eth0: mediatek frame engine at 0xbe100000, irq 21
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.445771] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.451437] mtk_soc_eth 1e100000.ethernet: generated random MAC address 06:2b:dc:a0:b4:6c
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.470895] mtk_soc_eth 1e100000.ethernet wan: mediatek frame engine at 0xbe100000, irq 21
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.489243] i2c_dev: i2c /dev entries driver
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.501198] NET: Registered PF_INET6 protocol family
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.513473] Segment Routing with IPv6
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.520979] In-situ OAM (IOAM) with IPv6
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.528934] NET: Registered PF_PACKET protocol family
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.539558] 8021q: 802.1Q VLAN Support v1.8
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.553739] mt7530 mdio-bus:1f: MT7530 adapts as multi-chip module
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.592269] mt7530 mdio-bus:1f: configuring for fixed/rgmii link mode
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.606584] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.612496] mt7530 mdio-bus:1f: Link is Up - 1Gbps/Full - flow control rx/tx
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.627346] mt7530 mdio-bus:1f lan4 (uninitialized): PHY [mt7530-0:00] driver [MediaTek MT7530 PHY] (irq=23)
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.648902] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.655408] mt7530 mdio-bus:1f lan3 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7530 PHY] (irq=24)
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.676634] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.683141] mt7530 mdio-bus:1f lan2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7530 PHY] (irq=25)
Tue Mar 28 10:59:19 2023 kern.err kernel: [    3.704194] It's not a pdev
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.710762] mt7530 mdio-bus:1f lan1 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7530 PHY] (irq=26)
Tue Mar 28 10:59:19 2023 kern.info kernel: [    3.732823] DSA: tree 0 setup
```
I saw you do a lot of work on nvmem. Could you please extend the patch `601-of_net-add-mac-address-ascii-support.patch` to apply the `mac-address-ascii` implementation to `net/core/of_net.c`?

https://github.com/gregkh/linux/blob/a1effab7a3a35a837dd9d2b974a1bc4939df1ad5/net/core/of_net.c#L77

aiamadeus committed:
generic: update nvmem cell mac-address-ascii support

Instead of use mac-address-ascii in nvmem_get_mac_address
function, move it into of_get_mac_addr_nvmem function to
support more devices.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

DragonBluep commented:
Thanks, build run and confirm it works now.

Ansuel commented:
@aiamadeus thanks for the patch can you refresh it?

aiamadeus commented:
> thanks for the patch can you refresh it?

Refresh, it should be ok now.

csharper2005 commented:
@aiamadeus @Ansuel @DragonBluep good news. Thank you! 

P.S. I did the same for TP-Link EC330-G5u v1 in this PR - https://github.com/openwrt/openwrt/pull/12445
